core/src/main/java/org/infinispan/persistence/sifs/Index.java

```diff
@@ -77,18 +82,44 @@ public Index(NonBlockingManager nonBlockingManager, FileProvider fileProvider, P
       this.maxNodeSize = maxNodeSize;
       indexDir.toFile().mkdirs();
 
+      // If there are more index files than our segments, that means we can't load them
+      boolean hasExtraIndexFile = fileExists(segments);
+
       this.segments = new Segment[segments];
       this.flowableProcessors = new FlowableProcessor[segments];
       for (int i = 0; i < segments; ++i) {
          UnicastProcessor<IndexRequest> flowableProcessor = UnicastProcessor.create();
-         Segment segment = new Segment(i, temporaryTable);
+         Segment segment = new Segment(i, temporaryTable, !hasExtraIndexFile);
 
          this.segments[i] = segment;
          // It is possible to write from multiple threads
          this.flowableProcessors[i] = flowableProcessor.toSerialized();
       }
    }
 
+   boolean fileExists(int id) {
+      File indexFileFile = new File(indexDir.toFile(), "index." + id);
+      return indexFileFile.exists();
+   }
+
+   public static byte[] toIndexKey(int cacheSegment, org.infinispan.commons.io.ByteBuffer buffer) {
+      int segmentBytes = UnsignedNumeric.sizeUnsignedInt(cacheSegment);
+      byte[] indexKey = new byte[buffer.getLength() + segmentBytes];
+      UnsignedNumeric.writeUnsignedInt(indexKey, 0, cacheSegment);
+      System.arraycopy(buffer.getBuf(), buffer.getOffset(), indexKey, segmentBytes, buffer.getLength());
+
+      return indexKey;
+   }
+
+   static byte[] toIndexKey(int cacheSegment, byte[] bytes) {
```

Comment:
nit you could wrap the  `bytes` into `org.infinispan.commons.io.ByteBuffer` to avoid duplicate code.

Reply:
I am not a fan of allocting a ByteBuffer for no reason. However, I can do the inverse, change this method to take offset and length and then have the ByteBuffer method just invoke this one.